### PR TITLE
*: fix some leak problems

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve? 
Currently, we have some goroutine leak problems, such as TTL cache. 

### What is changed and how it works?
This PR uses the context to control the lifetime of these goroutines to avoid the potential leak problems.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: server/schedule/operator_controller.go [hunk 1]
```diff
@@ -72,21 +74,27 @@ type OperatorController struct {
...
-		opRecords:       NewOperatorRecords(),
+		opRecords:       NewOperatorRecords(ctx),
 		storesLimit:     make(map[uint64]*ratelimit.Bucket),
 		wop:             NewRandBuckets(),
 		wopStatus:       NewWaitingOperatorStatus(),
 		opNotifierQueue: make(operatorQueue, 0),
 	}
 }
 
+// Ctx returns a context.
```
Comment (disksing) [inline server/schedule/operator_controller.go]: I think we can make this comment more meaningful 😸

File: server/config/option.go [hunk 2]
```diff
@@ -299,12 +300,12 @@ func (o *ScheduleOption) AddSchedulerCfg(tp string, args []string) {
...
 
 // RemoveSchedulerCfg removes the scheduler configurations.
-func (o *ScheduleOption) RemoveSchedulerCfg(name string) error {
+func (o *ScheduleOption) RemoveSchedulerCfg(ctx context.Context, name string) error {
 	c := o.Load()
 	v := c.Clone()
 	for i, schedulerCfg := range v.Schedulers {
 		// To create a temporary scheduler is just used to get scheduler's name
-		tmp, err := schedule.CreateScheduler(schedulerCfg.Type, schedule.NewOperatorController(nil, nil), schedulerCfg.Args...)
+		tmp, err := schedule.CreateScheduler(schedulerCfg.Type, schedule.NewOperatorController(ctx, nil, nil), schedulerCfg.Args...)
```
Comment (nolouch) [inline server/config/option.go]: seem no necessary to pass a `ctx`,  just create a temporary schedule. Can we create a context in this function?

Comment (rleungx) [reply server/config/option.go]: In fact, e.g., we are trying to remove the balance leader scheduler, the TTL cache will leak if we create a new context here.

File: server/core/region_storage.go [hunk 3]
```diff
@@ -31,14 +31,14 @@ var dirtyFlushTick = time.Second
...
-	flushTime    time.Time
-	ctx          context.Context
-	cancel       context.CancelFunc
+	mu                  sync.RWMutex
+	batchRegions        map[string]*metapb.Region
+	batchSize           int
+	cacheSize           int
+	flushRate           time.Duration
+	flushTime           time.Time
+	regionStorageCtx    context.Context
```
Comment (shafreeck) [inline server/core/region_storage.go]: Why not keep the name `ctx` and `cancel`? It is a little verbose to use like this `regionStorage.regionStorageCtx`

File: server/heartbeat_streams.go [hunk 4]
```diff
@@ -39,26 +39,26 @@ type streamUpdate struct {
...
-	wg        sync.WaitGroup
-	ctx       context.Context
-	cancel    context.CancelFunc
-	clusterID uint64
-	streams   map[uint64]heartbeatStream
-	msgCh     chan *pdpb.RegionHeartbeatResponse
-	streamCh  chan streamUpdate
-	cluster   *RaftCluster
+	wg             sync.WaitGroup
+	hbStreamCtx    context.Context
```
Comment (shafreeck) [inline server/heartbeat_streams.go]: Same as above

**Comment by rleungx** [comment]
PTAL @disksing @nolouch @shafreeck

**Comment by disksing** [comment]
PTAL @nolouch

File: server/api/server_test.go [hunk 5]
```diff
@@ -71,6 +71,7 @@ func mustNewServer(c *C, opts ...func(cfg *config.Config)) (*server.Server, clea
 var zapLogOnce sync.Once
 
 func mustNewCluster(c *C, num int, opts ...func(cfg *config.Config)) ([]*config.Config, []*server.Server, cleanUpFunc) {
+	ctx, cleanup := context.WithCancel(context.Background())
```
Comment (lhy1024) [inline server/api/server_test.go]: would you like replace `cleanup` with `cancel`?

File: server/server.go [hunk 6]
```diff
@@ -295,26 +295,22 @@ func (s *Server) IsClosed() bool {
 	return atomic.LoadInt64(&s.isServing) == 0
 }
 
-var timeMonitorOnce sync.Once
-
 // Run runs the pd server.
 func (s *Server) Run(ctx context.Context) error {
-	timeMonitorOnce.Do(func() {
```
Comment (lhy1024) [inline server/server.go]: why remove `do once`?

File: server/server.go [hunk 7]
```diff
@@ -324,8 +320,8 @@ func (s *Server) Context() context.Context {
 	return s.serverLoopCtx
 }
 
-func (s *Server) startServerLoop() {
-	s.serverLoopCtx, s.serverLoopCancel = context.WithCancel(context.Background())
+func (s *Server) startServerLoop(ctx context.Context) {
+	s.serverLoopCtx, s.serverLoopCancel = context.WithCancel(ctx)
 	s.serverLoopWg.Add(3)
 	go s.leaderLoop()
```
Comment (lhy1024) [inline server/server.go]: do we need add ctx for them?

File: server/testutil.go [hunk 8]
```diff
@@ -39,21 +39,23 @@ import (
...
+	if err = s.Run(ctx); err != nil {
+		return nil, nil, err
 	}
 
 	cleanup := func() {
+		cancel()
 		s.Close()
 		testutil.CleanServer(cfg)
 	}
-	return cfg, s, cleanup, nil
```
[the remark this answers is not in the document]
Comment (rleungx) [reply server/testutil.go]: It's not used.

File: server/server.go (hunk 7 above)
Comment (rleungx) [reply server/server.go]: These three goroutines have their own contexts based on `serverLoopCtx`

File: server/server.go (hunk 6 above)
Comment (rleungx) [reply server/server.go]: After the context is canceled, the goroutine will exit. It seems no need to use it?

**Pull request merged**